classes/user_exception_selector.php

```diff
@@ -0,0 +1,193 @@
+     */
+    public function __construct($name, $options) {
+        $this->courseid = $options['courseid'];
+        $this->skipauth = $options['skipauth'];
+        parent::__construct($name, $options);
+    }
+
+    /**
+     * Returns users for whom exceptions can be set/unset in the associated course.
+     *
+     * This method is used both when getting the list of choices to display to the user, and also when validating a list of users
+     * that was selected.
+     *
+     * @param string $search the search string.
+     * @return object[][] An array of arrays of users. The outer array will only have one element with an option group name as key.
+     *                    The keys of the inner array will be user IDs and the values will be the corresponding user objects
+     *                    containing the fields returned by the method {@see required_fields_sql}.
+     *                    If no users match the provided `$search` string, an empty array is returned instead (without nesting).
+     * @throws coding_exception
+     * @throws dml_exception
+     */
+    public function find_users($search): array {
+        global $DB;
+        // By default, `wherecondition` retrieves all users except the deleted, not confirmed and guest.
+        [$wherecondition, $params] = $this->search_sql($search, 'u');
+        $params['courseid'] = $this->courseid;
+        // Different queries depending on whether we want to search among users with or without existing exceptions in the course.
+        if ($this->skipauth) {
+            $stringkey = 'users_with_exception';
+            $ejoin = 'JOIN';
+        } else {
+            $stringkey = 'users_without_exception';
+            $ejoin = 'LEFT JOIN';
+            $wherecondition .= ' AND e.id IS NULL';
+        }
+        // We write out the query including the `SELECT` clause once here to enable IDE insights via connected DB table definitions.
+        $countsql = "SELECT COUNT(1)
+                       FROM {user} u
+                       JOIN {user_enrolments} AS ue ON ue.userid = u.id
+                       JOIN {enrol} AS en ON ue.enrolid = en.id
+                       JOIN {course} AS course ON en.courseid = course.id
+                     $ejoin {availability_shibboleth2fa_e} e ON (e.userid = u.id AND e.courseid = course.id AND e.skipauth = 1)
+                      WHERE $wherecondition
+                            AND course.id = :courseid";
+        if (!$this->is_validating()) {
+            $userscount = $DB->count_records_sql($countsql, $params);
+            if ($userscount > $this->maxusersperpage) {
+                return $this->too_many_results($search, $userscount);
+            }
+        }
+        $sql = preg_replace(
+            pattern: '/^SELECT COUNT\(1\)/',
+            replacement: "SELECT {$this->required_fields_sql('u')}",
+            subject: $countsql,
+            limit: 1,
+        );
```

Comment:
Instead of replacing the SELECT clause, please create a function to build the SQL query (e.g. `build_find_users_query(string $search, bool $select, string $orderby): string`.

Reply:
Maybe you missed it, but as alluded to by my comment above the `SELECT COUNT` query, if the string actually starts with a `SELECT` clause, PHPStorm recognizes the SQL and gives insights into the tables mentioned, if a database connection is set up. I thought this might be more helpful in practice than dealing with string concatenation to construct the SQL query. What do you think about that? Is that too "niche" of a benefit?

Even though I don't particularly like the replacement, I though it was simple enough and once the Regex is compiled, it should be very fast. What would be the benefit of having a separate function? Seeing as it would not get called anywhere else.

Reply:
The disadvantage/danger that I see in this solution is that it depends on the query being formulated in this exact way. If someone changes the query to include a line break after the `SELECT`, or changes the `COUNT(1)` to `COUNT(u.id)` (for whichever reason), the replacing will not work anymore. It will even fail silently.

Furthermore, the replacement `"SELECT {$this->required_fields_sql('u')}"` depends on the user table being available with the alias `u`. If the table alias is changed in the original query, this will break. 

I think both of these points are not obvious when glancing over the function and thus may pose a risk in the future. You could add some comments to document this, of course, but I am generally of the opinion that replacing strings in SQL code is bad practice. (I also don't like my initial implementation, which relied heavily on string concatenation.)

Anyway, my suggestion is a function like this:

```php
protected function build_find_users_query(string $search, bool $selectcount = false): array {
    // By default wherecondition retrieves all users except the deleted, not confirmed and guest.
    [$wherecondition, $params] = $this->search_sql($search, 'u');
    $params['courseid'] = $this->courseid;

    if ($selectcount) {
        $fields = 'COUNT(1)';
        $orderby = '';
    } else {
        $fields = $this->required_fields_sql('u');
        [$sort, $sortparams] = users_order_by_sql('u', $search, $this->accesscontext);
        $orderby = "ORDER BY $sort";
        $params = array_merge($params, $sortparams);
    }

    return [
        "SELECT $fields
           FROM {user} u
           JOIN {user_enrolments} AS ue ON ue.userid = u.id
           JOIN {enrol} AS en ON ue.enrolid = en.id
           JOIN {course} AS course ON en.courseid = course.id
           JOIN {availability_shibboleth2fa_e} e ON (e.userid = u.id AND e.courseid = course.id AND e.skipauth = 1)
          WHERE $wherecondition
                AND course.id = :courseid
          $orderby", $params];
}
```
which is called like this:
```php
[$countsql, $countparams] = $this->build_find_users_query($search, selectcount: true);
$userscount = $DB->count_records_sql($countsql, $countparams);
```
and this:
```php
[$sql, $params] = $this->build_find_users_query($search);
$users = $DB->get_records_sql($sql, array_merge($params, $params));
```
PhpStorm recognizes this query.

Reply:
Yes, this is better. I changed the code accordingly.